File: src/riscv/uart.zig
const std = @import("std");
const Bus = @import("bus.zig");
const Trap = @import("trap.zig");

/// The interrupt request of UART.
pub const IRQ: u64 = 10;

/// Recieve holding register (for input bytes).
pub const RHR = Bus.Mmio.Uart.base;
/// Transmit holding register (for output bytes).
pub const THR = Bus.Mmio.Uart.base;
/// Line control register.
pub const LCR = Bus.Mmio.Uart.base + 3;
/// Line status register.
pub const LSR = Bus.Mmio.Uart.base + 5;

/// The reciever (RX) bit.
pub const LSR_RX: u8 = 1;
/// The transmitter (TX) bit.
pub const LSR_TX: u8 = 1 << 5;

pub fn Uart(comptime reader_type: anytype, comptime writer_type: anytype) type {
    return struct {
        const Self = @This();

        uart: [Bus.Mmio.Uart.size]u8 = undefined,
        interrupting: std.atomic.Atomic(bool) = std.atomic.Atomic(bool).init(false),

        reader: reader_type,
        writer: writer_type,

        thread: std.Thread = undefined,
        mutex: std.Thread.Mutex = .{},
        cond: std.Thread.Condition = .{},

        pub fn init(reader: anytype, writer: anytype) !Self {
            var ret = Self{
                .reader = reader,
                .writer = writer,
            };

            ret.uart[LSR - Bus.Mmio.Uart.base] |= LSR_TX;
            ret.thread = try std.Thread.spawn(.{}, threadFn, .{&ret});

            return ret;
        }

        fn threadFn(self: *Self) void {
            while (true) {
                const byte = self.reader.reader().readByte() catch unreachable;

                self.mutex.lock();
                defer self.mutex.unlock();

                while ((self.uart[LSR - Bus.Mmio.Uart.base] & LSR_RX) == 1)
                    self.cond.wait(&self.mutex);

                self.uart[0] = byte;
                self.interrupting.store(true, .Release);

                self.uart[LSR - Bus.Mmio.Uart.base] |= LSR_RX;
            }
        }

        pub fn load(self: *Self, comptime T: type, address: u64) Trap.Exception!u64 {
            if (T != u8) return error.LoadAccessFault;

            self.mutex.lock();
            defer self.mutex.unlock();

            return switch (address) {
                RHR => blk: {
                    self.cond.broadcast();
                    self.uart[LSR - Bus.Mmio.Uart.base] &= ~LSR_RX;
                    break :blk @intCast(self.uart[RHR - Bus.Mmio.Uart.base]);
                },
                else => @intCast(self.uart[address - Bus.Mmio.Uart.base]),
            };
        }

        pub fn store(self: *Self, comptime T: type, address: u64, value: u64) Trap.Exception!void {
            if (T != u8) return error.StoreAMOAccessFault;

            switch (address) {
                THR => {
                    self.writer.writer().print("{c}", .{@as(u8, @truncate(value))}) catch return error.StoreAMOAccessFault;
                    self.writer.flush() catch return error.StoreAMOAccessFault;
                },
                else => self.uart[(address - Bus.Mmio.Uart.base)] = @as(u8, @truncate(value)),
            }
        }

        pub fn isInterrupting(self: *Self) bool {
            return self.interrupting.swap(false, .Acquire);
        }
    };
}

File: src/uart.rs
//! The uart module contains the implementation of a universal asynchronous receiver-transmitter
//! (UART). The device is 16550a UART, which is used in the QEMU virt machine.
//! See the spec: http://byterunner.com/16550.html

#![allow(dead_code)]

use std::io;
use std::io::prelude::*;
use std::sync::{
    atomic::{AtomicBool, Ordering},
    Arc, Condvar, Mutex,
};
use std::thread;

/// The address which UART starts, same as QEMU virt machine.
pub const UART_BASE: u64 = 0x10000000;
/// The size of UART.
pub const UART_SIZE: u64 = 0x100;
/// The interrupt request of UART.
pub const UART_IRQ: u64 = 10;

/// Receive holding register (for input bytes).
pub const UART_RHR: u64 = UART_BASE;
/// Transmit holding register (for output bytes).
pub const UART_THR: u64 = UART_BASE + 0;
/// Line control register.
pub const UART_LCR: u64 = UART_BASE + 3;
/// Line status register.
/// LSR BIT 0:
///     0 = no data in receive holding register or FIFO.
///     1 = data has been receive and saved in the receive holding register or FIFO.
/// LSR BIT 5:
///     0 = transmit holding register is full. 16550 will not accept any data for transmission.
///     1 = transmitter hold register (or FIFO) is empty. CPU can load the next character.
pub const UART_LSR: u64 = UART_BASE + 5;

/// The receiver (RX) bit.
pub const UART_LSR_RX: u8 = 1;
/// The transmitter (TX) bit.
pub const UART_LSR_TX: u8 = 1 << 5;

#[derive(Debug, Clone)]
pub struct Uart {
    /// Pair of an array for UART buffer and a conditional variable.
    uart: Arc<(Mutex<[u8; UART_SIZE as usize]>, Condvar)>,
    /// Bit if an interrupt happens.
    interrupting: Arc<AtomicBool>,
}

impl Uart {
    /// Create a new `Uart` object.
    pub fn new() -> Self {
        let uart = Arc::new((Mutex::new([0; UART_SIZE as usize]), Condvar::new()));
        let interrupting = Arc::new(AtomicBool::new(false));
        {
            let (uart, _cvar) = &*uart;
            let mut uart = uart.lock().expect("failed to get an UART object");
            // Transmitter hold register is empty.
            uart[(UART_LSR - UART_BASE) as usize] |= UART_LSR_TX;
        }

        let mut byte = [0; 1];
        let cloned_uart = uart.clone();
        let cloned_interrupting = interrupting.clone();
        let _uart_thread_for_read = thread::spawn(move || loop {
            match io::stdin().read(&mut byte) {
                Ok(_) => {
                    let (uart, cvar) = &*cloned_uart;
                    let mut uart = uart.lock().expect("failed to get an UART object");
                    // Wait for the thread to start up.
                    while (uart[(UART_LSR - UART_BASE) as usize] & UART_LSR_RX) == 1 {
                        uart = cvar.wait(uart).expect("the mutex is poisoned");
                    }
                    uart[0] = byte[0];
                    cloned_interrupting.store(true, Ordering::Release);
                    // Data has been receive.
                    uart[(UART_LSR - UART_BASE) as usize] |= UART_LSR_RX;
                }
                Err(e) => {
                    println!("{}", e);
                }
            }
        });
        Self { uart, interrupting }
    }

    /// Return true if an interrupt is pending. Clear the interrupting flag by swapping a value.
    pub fn is_interrupting(&self) -> bool {
        self.interrupting.swap(false, Ordering::Acquire)
    }

    pub fn load(&mut self, addr: u64) -> u8 {
        let (uart, cvar) = &*self.uart;
        let mut uart = uart.lock().expect("failed to get an UART object");
        match addr {
            UART_RHR => {
                cvar.notify_one();
                uart[(UART_LSR - UART_BASE) as usize] &= !UART_LSR_RX;
                uart[(UART_RHR - UART_BASE) as usize]
            }
            _ => uart[(addr - UART_BASE) as usize],
        }
    }

    pub fn store(&mut self, addr: u64, value: u8) {
        let (uart, _cvar) = &*self.uart;
        let mut uart = uart.lock().expect("failed to get an UART object");
        match addr {
            UART_THR => {
                print!("{}", value as char);
                io::stdout().flush().expect("failed to flush stdout");
            }
            _ => {
                uart[(addr - UART_BASE) as usize] = value;
            }
        }
    }
}